tests/rptest/utils/cluster_topology.py

```diff
@@ -0,0 +1,234 @@
+        self.nodes = []
+
+    def add_rack(self, rack: Rack):
+        assert rack.name not in self.racks, f"rack {rack.name} already exists in {self.name}"
+        self.racks[rack.name] = rack
+
+    def add_node(self, node):
+        self.nodes.append(node)
+
+    def __str__(self):
+        return f"region: {self.name}, nodes: [{','.join([n.account.hostname for n in self.nodes])}], racks: {[str(r) for _, r in self.racks.items()]}"
+
+
+class ClusterTopology():
+    UNASSIGNED_REGION = "__ducktape_unassigned"
+
+    def __init__(self, network_device='eth0'):
+        self.regions = {}
+        self.network_configuration = []
+        self.node_qdisc = []
+        self.network_device = network_device
+
+    def _unassigned_region(self):
+        return Region(ClusterTopology.UNASSIGNED_REGION)
+
+    def add_region(self, region: Region):
+        assert region not in self.regions, f"region {region} already exists in cluster"
+        self.regions[region.name] = region
+
+    def add_rack(self, rack: Rack):
+        if ClusterTopology.UNASSIGNED_REGION not in self.regions:
+            self.add_region(Region(ClusterTopology.UNASSIGNED_REGION))
+
+        self.regions[ClusterTopology.UNASSIGNED_REGION].add_rack(rack)
+
+    def add_rack(self, region_name, rack: Rack):
+        assert region_name in self.regions, f"can not add rack to region {region_name} that does not exists"
+        self.regions[region_name].add_rack(rack)
+
+    def get_region(self, region_name: str):
+        return self.regions[region_name]
+
+    def add_node(self, region_name, rack_name, node):
+        assert region_name in self.regions, "can not add node to region that does not exists"
+        region: Region = self.regions[region_name]
+
+        assert rack_name in region.racks, f"can not add node to the rack {rack_name} that does not exists in {region_name} region"
+        region.racks[rack_name].add_node(node)
+
+    def nodes_in(self, region=None, rack=None, recursive=True):
+        if region is None and ClusterTopology.UNASSIGNED_REGION in self.regions:
+            if rack in self.regions[ClusterTopology.UNASSIGNED_REGION].racks:
+                return self.regions[
+                    ClusterTopology.UNASSIGNED_REGION].racks[rack].nodes
+        if region in self.regions:
+            rg = self.regions[region]
+            if rack is not None and rack in rg.racks:
+                return rg.racks[rack].nodes
+
+            if not recursive:
+                return rg.nodes
+            else:
+                return sum([r.nodes for r in rg.racks.values()], [])
+
+    def _ip_address(self, node):
+        res = node.account.ssh_output(f'hostname -i')
+        return res.strip().decode('utf-8')
+
+    def add_connection_spec(self, spec: TopologyConnectionSpec):
+        self.network_configuration.append(spec)
+
+    def start_with_topology(self, redpanda: RedpandaService, **kwargs):
+        for n in redpanda.nodes:
+            node_tn = self.placement_of(n)
```

Comment:
nit: guess we should `assert node_tn` because the caller is expected to `add_node` for every broker (or pick a random placement?).

Reply:
we may have nodes that aren't assigned anywhere

Reply:
we may have nodes that aren't assigned anywhere